build/BuildFunctions.psm1

```diff
@@ -53,77 +40,30 @@ function Initialize-PsTask([string]$task) {
     }
 }
 
-function Initialize-TsTask([string]$task) {
-    Write-Host "Initialize TypeScript task $task"
-    pushd $task
-    try {
-        if (-not (Test-Path node_modules)) {
-            npm install --loglevel error | Out-Null
-        }
-    } finally {
-        popd
-    }
-}
+function Invoke-AllMochaTests([string[]]$tasks, [string]$reporter, [switch]$throwOnError) {
+    Write-Host "Testing Tasks"
 
-# Compile functions should be able to be replaced by MSBuild.
-function Invoke-CompileAll([string[]]$tasks) {
-    Write-Host "Compiling TypeScript modules and tasks"
-    Invoke-CompileCommon
     $jobs = $tasks | % {
         Start-Job {
             cd $using:pwd
-            Import-Module ./build/BuildFunctions.psm1
-            Invoke-CompileTask -task $using:_
-        }
-    }
-    $jobs | Wait-Job | Receive-Job -Wait -AutoRemoveJob
-}
-
-function Invoke-CompileCommon() {
-    Write-Host "Compiling common modules"
-    pushd common
-    try {
-        Write-Verbose "Running: tsc"
-        tsc
-        if ($LASTEXITCODE -ne 0) {
-            throw "tsc failed for common modules"
+            Import-Module ./build/BuildFunctions.psm1 -ArgumentList $using:VerbosePreference
+            Invoke-MochaTest -task $using:_ -reporter $using:reporter
         }
-    } finally {
-        popd
     }
-}
 
-function Invoke-CompileTask($task) {
-    Write-Host "Compiling task $task"
-    pushd $task
-    try {
-        # Common module changes as part of development.
-        # Update the common module every time.
-        npm install ../common --loglevel warn | Out-Null
-
-        Write-Verbose "Running: tsc"
-        tsc
-        if ($LASTEXITCODE -ne 0) {
-            throw "tsc failed for task $task"
+    $allJobErrors = @()
+    do {
+        if ($jobs.Count -gt 1) {
+            $jobs | Wait-Job -Any | Receive-Job -Wait -AutoRemoveJob -ErrorVariable jobErrors
```

Comment:
Do we need `Wait-Job` if we already have `Receive-Job -Wait`?

Reply:
`Receive-Job -Wait` will output to the console the output of all jobs, interlacing the output. `Wait-Job -Any` will return when a single job is complete. This combination allows us to display test results to the console as quickly as they are available, but in single understandable chunks.